src/network/namespacemanager.cpp

```diff
@@ -0,0 +1,149 @@
+#include "network/namespacemanager.hpp"
+
+namespace aos::common::network {
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+Error NamespaceManager::Init(sm::networkmanager::NetworkInterfaceManagerItf& netIf)
+{
+    LOG_DBG() << "Init namespace manager";
+
+    mNetIf = &netIf;
+
+    if (!std::filesystem::exists(cPathToNetNs)) {
+        if (!std::filesystem::create_directory(cPathToNetNs)) {
+            return Error(ErrorEnum::eFailed, "failed to create network namespace directory");
+        }
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error NamespaceManager::CreateNetworkNamespace(const String& instanceID)
+{
+    LOG_DBG() << "Create network namespace: instanceID=" << instanceID;
+
+    auto path = std::filesystem::path(cPathToNetNs) / instanceID.CStr();
+
+    if (std::filesystem::exists(path)) {
+        return ErrorEnum::eNone;
+    }
+
+    Error                err;
+    StaticString<cNSLen> originalNsPath;
+
+    if (err = originalNsPath.Format(cNSPathFormat, getpid(), gettid()); !err.IsNone()) {
+        return err;
+    }
+
+    int originalNsFd = open(originalNsPath.CStr(), O_RDONLY | O_CLOEXEC);
+    if (originalNsFd < 0) {
+        return Error(ErrorEnum::eFailed, strerror(errno));
+    }
+
+    auto closeFd         = [](int* fd) { close(*fd); };
+    auto originalNsFdPtr = std::unique_ptr<int, decltype(closeFd)>(&originalNsFd, closeFd);
+
+    // Unshare to create the new network namespace
+    if (unshare(CLONE_NEWNET) != 0) {
+        return Error(ErrorEnum::eFailed, strerror(errno));
+    }
+
+    int nsFd = open(path.c_str(), O_CREAT | O_EXCL | O_RDWR, 0444);
+    if (nsFd < 0) {
+        return Error(ErrorEnum::eFailed, strerror(errno));
+    }
+
+    close(nsFd);
+
+    auto autoSwitchToOriginNs = DeferRelease(&originalNsFd, [&path, &err](int* fd) {
+        if (setns(*fd, CLONE_NEWNET) != 0) {
+            LOG_ERR() << "Failed to return to original namespace: " << strerror(errno);
+
+            if (unlink(path.c_str()) != 0) {
+                LOG_ERR() << "Failed to remove namespace file: " << strerror(errno);
+            }
+
+            return;
+        }
+
+        if (!err.IsNone()) {
+            LOG_ERR() << "Error in creating network namespace, err=" << err;
+
+            if (unlink(path.c_str()) != 0) {
+                LOG_ERR() << "Failed to remove namespace file: " << strerror(errno);
+            }
+        }
+    });
+
+    StaticString<cNSLen> currentNsPath;
+
+    if (err = currentNsPath.Format(cNSPathFormat, getpid(), gettid()); !err.IsNone()) {
+        return err;
+    }
+
+    if (mount(currentNsPath.CStr(), path.c_str(), "none", MS_BIND, nullptr) != 0) {
+        return Error(ErrorEnum::eFailed, strerror(errno));
+    }
+
+    if (err = mNetIf->BringUpInterface("lo"); !err.IsNone()) {
```

Comment:
I don't see delete interface call. Is it by design?

Reply:
What do you mean by delete interface call? if you mean deleting `lo` interface then this handles the OS namespace itself, not only `lo` interface but all resources that were allocated network namespace. and `lo` interface is created namespace itself, but in our case needs to up, allow the use of localhost functionality for example.